fix(links) fix hydration overriding wrong elements of the ssr html

EugeneHlushko opened the pull request:
Closes #3287

hydration with index based keys maps wrong data to ssred html elements, wrong indexes come from printable.md being in the content during ssg config run.

montogeek commented:
Please try to keep formatting changes out of normal PRs.
Also, I don't get issue error at all right now in production. Did you reproduce it?

ejez commented:
Issue still in production and also after this PR:

![Screenshot from 2019-09-20 11-20-32](https://user-images.githubusercontent.com/48388358/65312301-e4b4a000-db9a-11e9-859a-039edd7e7ebd.png)

EugeneHlushko commented:
@ejez i cannot reproduce on netlife preview nor locally on this branch :O

ejez commented:
Delete browser cache/history related to the site, then open guides main url, reload it once or twice, then hover over a sub topic below "asset management".

Issue is not present when already on a sub topic url.

Try different devices/browsers

ejez commented:
(do not click on the subtopic links, just hover. clicking will take you to the correct suptopic, but hovering on or inspecting the sidebar item shows incorrect link)

EugeneHlushko commented:
@ejez i cannot reproduce with cache disabled, in incognito, etc. @montogeek its your call now

ejez commented:
![webpack org js-incorrect-links-issue](https://user-images.githubusercontent.com/48388358/65532093-e1efdd00-df03-11e9-9f1b-b37058376b45.gif)

firefox 69.0 / ubuntu 18.04

ejez commented:
Also when in a subtopic page and after refreshing the page, the subtree will show a different subtopic open:

![Screenshot from 2019-09-25 17-52-06](https://user-images.githubusercontent.com/48388358/65613402-a403bf00-dfbe-11e9-8e29-8fa8cf1b5bef.png)


![Screenshot from 2019-09-25 17-58-55](https://user-images.githubusercontent.com/48388358/65613447-bda50680-dfbe-11e9-83e7-7a8272459f86.png)

ejez commented:
@EugeneHlushko you can reproduce here:
https://www.browserling.com/browse/win/7/firefox/68/https%3A%2F%2Fwebpack.js.org%2Fguides%2F

and here:
https://www.browserling.com/browse/win/7/firefox/68/https%3A%2F%2Fdeploy-preview-3288--webpackjsorg-netlify.netlify.com%2Fguides%2F

EugeneHlushko commented:
Yup its reproducible in that firefox, will take a look into it again, ty @ejez

EugeneHlushko commented:
Interestingly i cannot reproduce it in latest firefox either ( 69.0.2 (64-bit) )

montogeek commented:
I just installed a new Firefox channel, which is basically a fresh installation and it is reproducible.

<img width="1440" alt="Screenshot 2019-10-10 at 07 43 21" src="https://user-images.githubusercontent.com/1002461/66569695-b7239c80-eb31-11e9-8f33-1002736bfa0c.png">


It is also reproducible in this PR Netflix preview deploy. As you can see in above image, "Output management" is missing from the server rendering, it is hydrated by React client side. That gives us 2 ideas:
1. Why is "Output management" not rendering in the server side like the rest of the links? This would be the correct fix.
2. What are we missing when building the app that is creating this _mismatch_ which is confusing React hydration?

ejez commented:
made a local test:

the issue disappears if we set `mode: 'none',` or `mode: 'development',` in `webpack.prod.js`

The issue might be narrowed down by checking individually the plugins enabled by `mode: 'production',` or setting `process.env.NODE_ENV` to `development`

will try to check later

ejez commented:
The issue disappears if we set `process.env.NODE_ENV` to `development` (while keeping `mode: 'production'`)

ejez commented:
While waiting to identify the issue, we can get the site working properly by disabling ssr for the sidebar,
I am not familiar with "react" but I tested the following and it is working:
https://deploy-preview-3314--webpackjsorg-netlify.netlify.com/guides/development/
PR #3314 

`src/components/Site/Site.jsx` :
```diff
 class Site extends React.Component {
   state = {
+    hasMounted: false,
     mobileSidebarOpen: false
   };
 
+  componentDidMount() {
+    this.setState({ hasMounted: true });
+  }
+
   render() {
     let { location } = this.props;
     let { mobileSidebarOpen } = this.state;

...

                         return (
                           <React.Fragment>
                             <Sponsors />
+                            { this.state.hasMounted ? (
                            <Sidebar
                              className="site__sidebar"
                              currentPage={location.pathname}
                              pages={this._strip(
                                section
                                  ? section.children
                                  : Content.children.filter(item => item.type !== 'directory' && item.url !== '/')
                              )}
                            />
+                            ) : '' }
                             <Page {...page} content={content} />
                             <Gitter />
                           </React.Fragment>
```

EugeneHlushko commented:
Disabling SSR for the sidebar will remove the links, this will mess up Google ranking of the pages and established internal websites linking.
I will look into what you guys found

EugeneHlushko commented:
@montogeek it looks like Output management is in SSR's HTML, trying to understand why its not visible, maybe HTML is invalid?

EugeneHlushko commented:
It is now fixed (writing this before preview lands so 🤞 )

I found that sort: 0 was causing a weird issue:
![after sort with 0](https://user-images.githubusercontent.com/10549495/66702699-26a4b200-ed13-11e9-9d74-9a307eafa680.png)

When we had sort: 0 item sort was behaving strange...

After i changed it to -1:

![after sort with -1](https://user-images.githubusercontent.com/10549495/66702704-31f7dd80-ed13-11e9-915a-0d90b14bb294.png)

EugeneHlushko commented:
For details i suggest you see last commit

EugeneHlushko commented:
Yup its fixed: https://www.browserling.com/browse/win/7/firefox/68/https%3A%2F%2Fdeploy-preview-3288--webpackjsorg-netlify.netlify.com%2Fguides%2F

@montogeek or @ejez please confirm

ejez commented:
Working! thanks @EugeneHlushko

montogeek commented on src/scripts/concatenate-docs.js:
```diff
@@ -49,7 +49,16 @@ function getDirectoryRecursive(basePath) {
-		const targetFile = fs.createWriteStream(targetFilePath);
+    const targetFile = fs.createWriteStream(targetFilePath);
+
+    targetFile.write(`---
```
It this needed? If so, why?

EugeneHlushko replied:
So this makes sure even the `printable.md` files have sort. Without title and contributors it fails to parse so i've added generic ones as they are harmless to printable files

montogeek commented on src/utilities/content-tree-enhancers.js:
```diff
@@ -76,6 +76,7 @@ function restructure(item, options) {
   enhance(item, options);
 
   if (item.children) {
+    // item.children.push(...item.children.splice(item.children.findIndex(child => child.name === 'printable.md'), 1));
```
Please explain

EugeneHlushko replied:
this is redundant, cleaning up 
```suggestion
```

montogeek commented:
> When we had sort: 0 item sort was behaving strange...

Why exactly? Something related with 0 array indexes?

Also from your screenshots, `21` is missing before but present in the second screenshot. Maybe another fix will be to guarantee the consecutive order of the sort property?

montogeek commented:
Also wonder if this is the same in other sections of the website.

EugeneHlushko commented:
Yeah so i've tested all sections as well which had 0 on sort, checking if its working is quite easy just hovering over links like we did to reproduce this one

EugeneHlushko commented:
Why sort failed, i am not sure, we have interesting sort function which is pretty straightforward, but for some reason it only worked (without my fix) if i did .sort(sort) twice in a row [here](https://github.com/webpack/webpack.js.org/blob/master/src/utilities/content-tree-enhancers.js#L82)... After fix 1 sort is enough.

montogeek commented:
Thanks!

EugeneHlushko commented:
🎉 finally

EugeneHlushko commented:
i think this was third tracked attempt to track down the issue, although the culprit is react's hydration that fails when sort order changes (there is an issue in react repo about this linked in previous pr)
